MdeModulePkg/Library/MemoryProtectionHobLib/DxeMemoryProtectionHobLib.c

```diff
@@ -0,0 +1,302 @@
+      }
+
+      if (gDxeMps.HeapGuardPoolType.Fields.EfiReservedMemoryType != 0) {
+        DEBUG ((
+          DEBUG_WARN,
+          "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Pool Guard set on \
+          EfiReservedMemoryType. This is not supported by Heap Guard system, disabling.\n",
+          __func__
+          ));
+        gDxeMps.HeapGuardPoolType.Fields.EfiReservedMemoryType = 0;
+      }
+
+      if (gDxeMps.HeapGuardPoolType.Fields.EfiRuntimeServicesCode != 0) {
+        DEBUG ((
+          DEBUG_WARN,
+          "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Pool Guard set on \
+          EfiRuntimeServicesCode. This is not supported by Heap Guard system, disabling.\n",
+          __func__
+          ));
+        gDxeMps.HeapGuardPoolType.Fields.EfiRuntimeServicesCode = 0;
+      }
+
+      if (gDxeMps.HeapGuardPoolType.Fields.EfiRuntimeServicesData != 0) {
+        DEBUG ((
+          DEBUG_WARN,
+          "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Pool Guard set on \
+          EfiRuntimeServicesData. This is not supported by Heap Guard system, disabling.\n",
+          __func__
+          ));
+        gDxeMps.HeapGuardPoolType.Fields.EfiRuntimeServicesData = 0;
+      }
+    }
+  }
+}
+
+/**
+  Populates gDxeMps global with the data present in the HOB. If the HOB entry does not exist,
+  this constructor will zero the memory protection settings.
+
+  @retval EFI_SUCCESS   The constructor always returns EFI_SUCCESS.
+**/
+EFI_STATUS
+EFIAPI
+DxeMemoryProtectionHobLibConstructor (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  VOID  *Ptr;
+
+  Ptr = GetFirstGuidHob (&gDxeMemoryProtectionSettingsGuid);
+
+  //
+  // Cache the Memory Protection Settings HOB entry
+  //
+  if (Ptr != NULL) {
+    if (*((UINT8 *)GET_GUID_HOB_DATA (Ptr)) != (UINT8)DXE_MEMORY_PROTECTION_SETTINGS_CURRENT_VERSION) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "\nThe version number of the DXE Memory Protection Settings HOB is invalid! Expected: %d, Actual: %d\n",
+        DXE_MEMORY_PROTECTION_SETTINGS_CURRENT_VERSION,
+        *((UINT8 *)GET_GUID_HOB_DATA (Ptr))
+        ));
+      DEBUG ((DEBUG_ERROR, "This usually happens when the Memory Protection Settings version was incremented\n"));
+      DEBUG ((DEBUG_ERROR, "and all modules have not been rebuilt with the new version number. Less likely but\n"));
+      DEBUG ((DEBUG_ERROR, "also possible is the HOB entry was corrupted or the producer of the HOB entry\n"));
+      DEBUG ((DEBUG_ERROR, "did not set the StructVersion field to DXE_MEMORY_PROTECTION_SETTINGS_CURRENT_VERSION.\n"));
+      ASSERT (FALSE);
+      ZeroMem (&gDxeMps, sizeof (gDxeMps));
+      return EFI_SUCCESS;
+    }
+
+    CopyMem (&gDxeMps, GET_GUID_HOB_DATA (Ptr), sizeof (DXE_MEMORY_PROTECTION_SETTINGS));
+    DxeMemoryProtectionSettingsConsistencyCheck ();
+  } else {
+    DEBUG ((

```

Comment:
Is this something we expect to tighten up soon (zeroed memory protection config if HOB not found)? If so, an inflection point like 2405 might make sense to do so.

Reply:
That's a good idea. I would lean towards defaulting to ship mode protections at a minimum here. I'll talk with @spbrogan.

Reply:
I chatted with Sean, he thought this was better to target in a future iteration than now.